drivers/pinctrl/pinctrl_npcx.c

```diff
@@ -0,0 +1,142 @@
+		group == 0x0d || (group >= 0x0f && group <= 0x12)) {
+		return true;
+	}
+#endif
+	return false;
+}
+
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, int func,
+	bool is_locked)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+	uint8_t alt_mask = BIT(alt->bit);
+
+	/*
+	 * func == 0 means select GPIO, otherwise Alternate Func.
+	 * inverted == 0:
+	 *    Set devalt bit to select Alternate Func.
+	 * inverted == 1:
+	 *    Clear devalt bit to select Alternate Func.
+	 */
+	if (!!func != alt->inverted) {
+		NPCX_DEVALT(scfg_base, alt->group) |=  alt_mask;
+	} else {
+		NPCX_DEVALT(scfg_base, alt->group) &= ~alt_mask;
+	}
+
+	if (is_locked && npcx_periph_pinmux_has_lock(alt->group)) {
+		NPCX_DEVALT_LK(scfg_base, alt->group) |= alt_mask;
+	}
+}
+
+static void npcx_periph_pupd_configure(const struct npcx_periph *pupd,
+	enum npcx_io_bias_type type)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+
+	if (type == NPCX_BIAS_TYPE_NONE) {
+		NPCX_PUPD_EN(scfg_base, pupd->group) &= ~BIT(pupd->bit);
+	} else {
+		NPCX_PUPD_EN(scfg_base, pupd->group) |=  BIT(pupd->bit);
+	}
+}
+
+static void npcx_periph_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PINMUX) {
+		/* Configure peripheral device's pinmux functionality */
+		npcx_periph_pinmux_configure(&pin->cfg.periph, 1,
+					     pin->flags.pinmux_lock);
+	} else if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PUPD) {
+		/* Configure peripheral device's internal PU/PD */
+		npcx_periph_pupd_configure(&pin->cfg.periph,
+			pin->flags.io_bias_type);
+	}
+}
+
+/* Pin-control local functions for IO pads */
+static void npcx_iopad_1p8_input_configure(const struct npcx_io_pad *pad)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+
+	NPCX_LV_GPIO_CTL(scfg_base, pad->lvol_ctrl) |= BIT(pad->lvol_bit);
+}
+
+static void npcx_io_pad_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->flags.type == NPCX_PINCTRL_TYPE_IO_PAD) {
+		if (pin->flags.io_1p8v_input) {
+			npcx_iopad_1p8_input_configure(&pin->cfg.io_pad);
+		}
+		/**
+		 * TODO: Add more io-pad configurations here.
+		 */
+	}
+}
+
+/* Pinctrl API implementation */
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+			   uintptr_t reg)
+{
+	ARG_UNUSED(reg);
+
+	/* Configure all peripheral devices' properties here. */
+	for (uint8_t i = 0; i < pin_cnt; i++) {
+		npcx_periph_configure(&pins[i]);
+	}
+
+	return 0;
+}
+
+static int npcx_pinctrl_init(const struct device *dev)
+{
+	/* Configure all IO pads' properties here. */
+	for (int i = 0; i < ARRAY_SIZE(io_pad_conf); i++) {
+		npcx_io_pad_configure(&io_pad_conf[i]);
+	}
+
+	return 0;
+}
+SYS_INIT(npcx_pinctrl_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+BUILD_ASSERT(sizeof(pinctrl_soc_pin_t) == 4, "Unexpected pinctrl_soc_pin_t size");
+BUILD_ASSERT(sizeof(struct npcx_periph) == 2, "Unexpected npcx_periph size");
```

Comment:
why is this required?

Reply:
Just in case to prevent unexpected struct size if the compiler version is changed.

Reply:
still don't see why `pinctrl_soc_pin_t` needs to have a fixed size.

Reply:
On our customers` projects, we do much effort to shrink ram usage since limited resources. These lines give us a notification that the size of ```pinctrl_soc_pin_t``` is changed. We can find a way to save ram usage possibly.

Reply:
As a user of the NPCX9 chipset, I can confirm code size is ongoing concern.  Because pinctrl touches so many device instances, there is potentially a big impact if the structure size increases.

Reply:
I still see no reasons that justify such build assert, code size is not a valid reason IMO. `pinctrl_soc_pin_t` can be documented accordingly, instead.

Reply:
Okay. I will remove them and add a reminder in the comment of ```pinctrl_soc_pin_t```.

Reply:
Done.

Reply:
Hi @keith-zephyr, if you still have any concern about this change, please give your comment freely.